feat: Return tracking in PR/DN

marination opened the pull request:
Docs: https://github.com/frappe/erpnext_documentation/pull/154

## Transaction
<details><summary>Purchase Receipt</summary>
  
  - Returned Qty fields in Items Table
    ![Screenshot 2020-07-31 at 4 01 17 PM](https://user-images.githubusercontent.com/25857446/89026621-a7867600-d346-11ea-86df-2a6d83881e7c.png)

  - % Returned in parent doc
   ![Screenshot 2020-07-31 at 4 02 02 PM](https://user-images.githubusercontent.com/25857446/89026630-ace3c080-d346-11ea-8896-73dee8f42f81.png)

  - Status 'Return Issued' added on original PR if 100% returned. If partially returned , priority given to 'To Bill'
   ![Screenshot 2020-07-31 at 4 03 48 PM](https://user-images.githubusercontent.com/25857446/89026705-cbe25280-d346-11ea-8542-90feab42c6a8.png)

</details>

<details><summary>Delivery Note</summary>

  - Returned Qty fields in Items Table
    ![Screenshot 2020-07-31 at 8 14 34 PM](https://user-images.githubusercontent.com/25857446/89046119-2f31ac00-d36a-11ea-8e17-a10e8ef0b670.png)

  - % Returned in parent doc
    ![Screenshot 2020-07-31 at 8 15 02 PM](https://user-images.githubusercontent.com/25857446/89046149-36f15080-d36a-11ea-837c-efe4c0f6ef94.png)

  - Status 'Return Issued' added on original DN if 100% returned. If partially returned , priority given to 'To Bill'
    ![Screenshot 2020-07-31 at 8 15 33 PM](https://user-images.githubusercontent.com/25857446/89046186-42447c00-d36a-11ea-8a1b-8651635021c9.png)

</details>

<details><summary>Reports Changes</summary>

- Added Returned Amount to **Delivered** and **Purchased** Items to be Billed Reports.
- **Pending Amount = Total Amount - Billed Amount  - Returned Amount**. Essentially only non-returned items should be counted as pending
- If **Pending Amount** > 0, only then will the item be visible in the Report. So if a PR is 50 % Billed and 50% Returned, it won't be visible in the Report.

![Screenshot 2020-09-10 at 6 48 42 PM](https://user-images.githubusercontent.com/25857446/92734685-626a5080-f396-11ea-8425-7259dae95e51.png)

</details>

ToDo:
- [x]  Return for DN
- [x]  Tests for PR/DN
- [x]  Report Fixes (Received Items to be Billed, Delivered Items to be Billed)
- [x] Patch to set qty and % 
- [x]  Test to check Correct Mapping and Billing % Updation (bill after return and bill before return)

marination committed:
feat: Return tracking in PR/DN

marination committed:
Merge branch 'develop' into pr-dn-return

marination committed:
chore: Delivery Note Return status

marination committed:
Merge branch 'pr-dn-return' of https://github.com/marination/erpnext into pr-dn-return

marination committed:
fix: Codacy and repositioned Returned Qty in DN Item

marination committed:
Merge branch 'develop' into pr-dn-return

marination committed:
chore: Tests for Purchase Receipt

marination committed:
chore: Tests for Delivery Note

marination committed:
fix: Patch and test codacy

marination committed:
Merge branch 'develop' into pr-dn-return

marination committed:
chore: Fix Received/Delivered Items to Billed Logic

marination committed:
chore: Patch to set returned qty in PR and DN

marination committed:
fix: Use independent item for DN Test

marination committed:
Merge branch 'develop' into pr-dn-return

marination committed:
Merge branch 'develop' into pr-dn-return

marination committed:
Merge branch 'develop' into pr-dn-return

marination committed:
Merge branch 'develop' into pr-dn-return

rohitwaghchaure commented:
During test found below issues

1. Make purchase receipt for 3 quantity, make purchase return for 1 quantity and submit it. After that do one more purchase return entry, system shows default quantity as 3. It should show default quantity as 2 because 1 quantity already has returned.

1. Make delivery note for 3 quantity, and do sales return for 1 quantity. Then make sales invoice for 2 quantity and check delivery note status. It shows the status as "To Bill", should show the status as "Completed".

<img width="1229" alt="Screenshot 2020-10-28 at 11 20 27 AM" src="https://user-images.githubusercontent.com/8780500/97397627-b1048780-190f-11eb-904a-15ae42f5ff24.png">

marination committed:
Merge branch 'develop' into pr-dn-return

marination committed:
fix: Billing % Logic and Map Pending Qty only in PR and DN

- Billing % should consider unreturned amount as total
- While mapping to return doc, map unreturned amount
- Added field Received Qty in Stock UOM, to tally against Returned Qty in PR
- PR billing percentage updation custom function
- In patch set received qty in stock uom first, then update returned qty and billing

marination committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into pr-dn-return

marination committed:
chore: Avoid multiline string in Translation & remove print statement

marination committed:
chore: Tests

- Added test for mapping secnd return doc
- Added test for billing % of partially returned doc
- Handled PR with 0 billing amount

marination committed:
fix: Sider

marination commented:
Reports bit is already ported to v12 https://github.com/frappe/erpnext/pull/23804

marination committed:
Merge branch 'develop' into pr-dn-return

marination committed:
Merge branch 'develop' into pr-dn-return

marination committed:
Merge branch 'develop' into pr-dn-return

gavindsouza commented on erpnext/accounts/report/received_items_to_be_billed/received_items_to_be_billed.py:
```diff
@@ -14,11 +14,19 @@ def execute(filters=None):
-		_("Project") + ":Link/Project:120", _("Item Code") + ":Link/Item:120",
-		_("Amount") + ":Currency:100", _("Billed Amount") + ":Currency:100", _("Amount to Bill") + ":Currency:100",
-		_("Item Name") + "::120", _("Description") + "::120", _("Company") + ":Link/Company:120",
+		_("Purchase Receipt") + ":Link/Purchase Receipt:160",
```
Use JSON style API response

gavindsouza commented on erpnext/controllers/sales_and_purchase_return.py:
```diff
@@ -203,10 +203,46 @@ def get_already_returned_items(doc):
+			and parent.docstatus = 1
+			and parent.is_return = 1
+			and {2} = %(row_name)s
+		""".format(child_doctype, doctype, reference_field),
```
~`.format`~. Try using ORM instead.

gavindsouza commented on erpnext/controllers/sales_and_purchase_return.py:
```diff
@@ -203,10 +203,46 @@ def get_already_returned_items(doc):
+		},
+		as_dict=1)
+
+	return data[0]
```
`IndexError`?

nabinhait commented:
@marination resolve conflict

marination committed:
Merge branch 'develop' into pr-dn-return

marination committed:
Merge branch 'develop' into pr-dn-return

marination committed:
Merge branch 'develop' into pr-dn-return

marination committed:
chore: Use JSON style response and use ORM

- Use JSON style response for report columns
- Use ORM instead of frappe.db.sql
- Remove returned % from list view
